In [1]:
!pip install -U scipy

  Attempting uninstall: scipy
    Found existing installation: scipy 1.9.0
    Uninstalling scipy-1.9.0:
      Successfully uninstalled scipy-1.9.0


In [2]:
# import numpy and pandas
import pandas as pd
import numpy as np

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the San Francisco. We will start by loading the dataset and examining its contents. 

In [3]:
salary = pd.read_csv('C:/Users/mahna/Ironhack/Week4/Day3/lab-Hypothesis-Testing/data/Salaries.csv')


C:\Users\mahna\AppData\Local\Temp\ipykernel_12620\3630000816.py:1: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  salary = pd.read_csv('C:/Users/mahna/Ironhack/Week4/Day3/lab-Hypothesis-Testing/data/Salaries.csv')


Examine the `salaries` dataset using the `head` function below.

In [4]:
salary.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

We see from looking at the `head` function that there is quite a bit of missing data. Get the amount of missing data in every column

In [5]:
salary.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Get the shape of the dataframe

In [6]:
salary.shape

(148654, 13)

Given output of the previous two cells, drop the corresponding column and compute again the amount of missing values.

In [7]:
salaries = salary.copy()

In [8]:
salaries = salaries.drop(['Notes'], axis =1)

In [9]:
salaries.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
dtype: int64

Check out what are the possible values of the column "Status".

In [10]:
salaries['Status'].unique()

array([nan, 'PT', 'FT'], dtype=object)

In [11]:
salaries['Status'].value_counts(dropna=False)

NaN    110535
FT      22334
PT      15785
Name: Status, dtype: int64

Drop any row with missing values in the "Status" column and compute again the number of missing values.

In [12]:
salaries= salaries.dropna(subset = ['Status'])

In [13]:
salaries.shape

(38119, 12)

In [14]:
salaries.isna().sum()

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Check out the types of each column and see if they make sense.

In [15]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  object 
 4   OvertimePay       38119 non-null  object 
 5   OtherPay          38119 non-null  object 
 6   Benefits          38119 non-null  object 
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.8+ MB


In [16]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer   257340.0   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department   302068.0   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.0  342802.63  38780.04  471952.64         510732.68  2014   
110532    10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.0    82313.7  96570.66  339653.70         436224.36  2014   
110534         0.0   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.0   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  
110531  San Francisco     PT  
110532  San Francisco     FT  
110533  San Francisco     PT  
110534  San Francisco     FT  
110535  San Francisco     FT

Do any type conversions and reset the index.

In [18]:
df = salaries.reset_index()
del df['index']

In [19]:
df.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer   257340.0   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department   302068.0   

  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.0  342802.63  38780.04  471952.64         510732.68  2014   
1    10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.0    82313.7  96570.66  339653.70         436224.36  2014   
3         0.0   19266.72  91302.46  326716.76         418019.22  2014   
4         0.0   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  
0  San Francisco     PT  
1  San Francisco     FT  
2  San Francisco     PT  
3  San Francisco     FT  
4  San Francisco     FT

In [21]:
convert_dict = {'BasePay': float,
                'OvertimePay': float,
                'OtherPay':float,
                'Benefits':float
                }
df2 = df.astype(convert_dict)
#convert the numbers into float

In [23]:
df2.info() #now they are all float

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38119 entries, 0 to 38118
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  float64
 4   OvertimePay       38119 non-null  float64
 5   OtherPay          38119 non-null  float64
 6   Benefits          38119 non-null  float64
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(6), int64(2), object(4)
memory usage: 3.5+ MB


In [24]:
df2=df2.reset_index(drop=True)

In [25]:
df2.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  
0  San Francisco     PT  
1  San Francisco     FT  
2  San Francisco     PT  
3  San Francisco     FT  
4  San Francisco     FT

Check out if "TotalPayBenefits" = "BasePay" + "OvertimePay" + "OtherPay" + "Benefits"

In [19]:
df['tpb_total?'] = df["TotalPayBenefits"] == df["BasePay"] + df["OvertimePay"] + df["OtherPay"] + df["Benefits"]

In [20]:
df['tpb_total?']

0         True
1        False
2         True
3        False
4         True
         ...  
38114    False
38115    False
38116    False
38117    False
38118    False
Name: tpb_total?, Length: 38119, dtype: bool

What is the percetage of employees for which the previous assumption is not True?

In [21]:
df['tpb_total?'].value_counts(normalize = True).mul(100).astype(str)+'%'

False    61.625436134211284%
True     38.374563865788716%
Name: tpb_total?, dtype: object

There are different departments in the city. List all departments and the count of employees in each department.

In [22]:
df.groupby(['JobTitle']).agg({'Id': 'nunique'})
#pokemon.groupby(['Combo Type']).agg({'A/D Ratio':max}).reset_index().sort_values('A/D Ratio',ascending=False)

Id
JobTitle                          
ACPO,JuvP, Juv Prob (SFERS)      1
ASR Senior Office Specialist    22
ASR-Office Assistant            15
Account Clerk                   93
Accountant I                     2
...                             ..
Wire Rope Cable Maint Sprv       1
Worker's Comp Supervisor 1       6
Worker's Compensation Adjuster  26
X-Ray Laboratory Aide           35
Youth Comm Advisor               1

[996 rows x 1 columns]

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of **all FT workers is significantly different from $75/hr**. Get first the hourly wage by dividing "TotalPayBenefits" by 50 weeks (assuming 10 labour days of holidays) and by 40hrs (assuming a 40hrs week).

$$Hourly Wage = \frac{TotalPayBenefits}{1 year}\frac{1 year}{50 Week}\frac{1 Week}{40 hr}$$

Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [27]:
# Your code here: (compute the "Hourly_Wage")

In [28]:
import scipy.stats as st
import numpy as np

In [68]:
df2['HourlyWage'] = df2['TotalPayBenefits']/(50*40)

In [69]:
df2.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  HourlyWage  
0  San Francisco     PT  255.366340  
1  San Francisco     FT  239.826105  
2  San Francisco     PT  218.112180  
3  San Francisco     FT  209.009610  
4  San Francisco     FT  208.717550

In [27]:
# Your code here: (Compute the mean hourly wage for all the "FT" employees)

In [28]:
# def mymean(df):
#     if 'FT' in df['Status']:
#         return stat.mean(df["HourlyWage"])

In [31]:
ftwage= df.loc[df['Status']=='FT', 'HourlyWage']
sample_mean= ftwage.mean()
sample_mean

69.264206407271

In [32]:
len(df.loc[df['Status']=='FT', 'HourlyWage'])

22334

In [33]:
pop_mean= df['HourlyWage'].mean()
pop_mean

50.130719334189436

In [34]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "FT" employees
import statistics as stat
import math

# Computing the t - statistic for the sample mean:
t_statistic= (sample_mean - pop_mean) / (std /sqrt(n))
t_score = (hourly_wage_FT - 75)/(sample_sd / (math.sqrt(len(sample))))
degree of freedom(df)= len(sample)-1

In [33]:
import statistics as stat
import math

In [35]:
stat.stdev(ftwage)

23.939610982990395

In [36]:
# Computing the t - statistic for the sample mean:
t_statistic= (sample_mean - pop_mean) / (23.939610982990395 /math.sqrt(22334))


In [37]:
t_statistic

119.44288764548641

In [38]:
t_score = (69.264206407271 - 75)/(23.939610982990395 / (math.sqrt(len(ftwage))))

In [39]:
t_score

-35.80631941460793

In [39]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

In [40]:
import scipy.stats

In [41]:
upper_critical_value = scipy.stats.t.ppf(q=1-0.05/2, df=len(ftwage)-1)
lower_critical_value = scipy.stats.t.ppf(q=0.05/2, df=len(ftwage)-1)
#Remeber: if you use a onetail (left/rigth) or a two tail test

In [42]:
print(upper_critical_value)
print(lower_critical_value)

1.9600702128845593
-1.9600702128845597


For a two tail test; If(statistic > lcv) and (statistic<ucv) -> Accept H0
Here, the t-score(-35.80631941460788) < lower_critical_value. So, H0 rejected. Therefore, the population mean is different from the sample mean.

In [43]:
# Method 2: Use the p-value method.
# Your code here:

#for a two tail test:
p_value = (1-(scipy.stats.t.cdf(t_score, df = len(sample)-1)))*2
#for a right tail test:
p_value = 1 - (stats.t.cdf(t_score, df = len(police)-1))
#for left:
p_value = stats.t.cdf(t_score, df = len(police)-1))

In [43]:
p_value = (1-(scipy.stats.t.cdf(t_score, df = len(ftwage)-1)))*2

In [44]:
p_value

2.0

The p_value is > than 0.05, therefore, null hypothesis is rejected.

In [45]:
#Another way:
area_of_statistic = scipy.stats.t.cdf(t_score, df=len(ftwage)-1)
area_of_lcv= 0.05/2
area_of_ucv= (1-(0.05/2))


In [46]:
print(area_of_statistic)
print(area_of_lcv)
print(area_of_ucv)

2.2589156167156547e-273
0.025
0.975


For two tail test, if (area_of_statistic> 0.05/2) and (area_of_statistic<(1-0.05/2) -> Accept H0
Here, area_of_statistics>area_of_ucv, Reject H0

In [48]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

In [49]:
# t_stat, p_value = stats.ttest_1samp(sample, pop_mean)

In [55]:
t_stat, p_value = scipy.stats.ttest_1samp(ftwage, pop_mean)

print('The t_stat = {}'.format(t_stat))
print('The p_value = {}'.format(p_value))
print()
print('p_value>0.05---> We reject H0!--> Salary did increase!')

The t_stat = -536432.0252489669
The p_value = 0.0

p_value>0.05---> We reject H0!--> Salary did increase!


Are all the methods in agreement?

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are **higher than last year's mean of $86000/year for all salaried employees** (use the column "TotalPayBenefits". Test  hypothesis using a 95% confidence interval.

Hint: Use apply and a lambda function to check in "Police" is in the "JobTitle" to get all the "Police" jobs.

In [51]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "Police" employees

In [48]:
#Filter the Police
police = salaries[salaries['JobTitle'].str.contains('Police')]

In [49]:
#Compute all value needed for the test:
sample= police['TotalPayBenefits']
mean_salary= np.mean(police['TotalPayBenefits'])
std_salary = np.std(police['TotalPayBenefits'])
df = len(police)-1
pop_mean = 86000

In [50]:
#Compute the t score
t_score= (mean_salary - pop_mean) / (std_salary/ (math.sqrt(len(police))))
t_score

50.266207472174855

In [55]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

In [51]:
st.chi2.ppf(0.95,df=df)

2002.5198203441964

It means, to cover 95% values we have to reach 2002.52 value.

In [57]:
# Method 2: Use the p-value method.
# Your code here:

In [52]:
#twotailed test
p_value = (1-(scipy.stats.t.cdf(t_score, df = len(police)-1)))*2
p_value

0.0

In [53]:
#Another way:
area_of_statistic = scipy.stats.t.cdf(t_score, df=len(police)-1)
area_of_lcv= 0.05/2
area_of_ucv= (1-(0.05/2))

print(area_of_statistic)
print(area_of_lcv)
print(area_of_ucv)

1.0
0.025
0.975


For two tail test, if (area_of_statistic> 0.05/2) and (area_of_statistic<(1-0.05/2) -> Accept H0 Here, area_of_statistics>area_of_ucv, Reject H0

In [148]:
# Ho: there are no differences
# Ha: there are differences
# p_value = P(table | Ho) = P(table | there are no differences) = 0 < 0.05
#seems like the salaries are actually quite differentiated.

In [149]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

In [57]:
scipy.stats.ttest_1samp(sample,pop_mean)

Ttest_1sampResult(statistic=50.252984742109945, pvalue=0.0)

In [58]:
from scipy import stats
t_stat, p_value = scipy.stats.ttest_1samp(sample, pop_mean)

print('The t_stat = {}'.format(t_stat))
print('The p_value = {}'.format(p_value))
print()
print('p_value>0.5---> We reject H0!--> Salary did increase!')

The t_stat = 50.252984742109945
The p_value = 0.0

p_value>0.5---> We reject H0!--> Salary did increase!


The workers from the "JobTitle" with the most employees have complained that their hourly wage is **less than $35/hour**. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

H0 : Workers Salary <$35/hr
H1 : Workers Salary >=$35/hr

In [69]:
worker = salaries[salaries['JobTitle'].str.contains('Worker')]

H0: salary< $35
H1: salary >=35

In [87]:
# Your code here: (Get the department which has most employees)

dept_emp_num =  salaries.groupby('JobTitle')['Id'].count().nlargest(1)
print(dept_emp_num)

JobTitle
Transit Operator    2479
Name: Id, dtype: int64


In [59]:
# Your code here: (Get the department which has most employees)
#df2['JobTitle'].mode()
df2['JobTitle'].value_counts().idxmax()

'Transit Operator'

In [ ]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider the right "JobTitle" employees

In [71]:
trans= df2[df2['JobTitle']=='Transit Operator']

In [73]:
wage_trans=trans['HourlyWage']
mean_trans =wage_trans.mean()

In [74]:
from scipy.stats import sem

In [75]:
std_error_trans = sem(wage_trans)
t_trans = (mean_trans - 35) / std_error_trans
t_trans

19.064990324906578

19.06499032490655

In [76]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
wage_trans.shape

(2479,)

In [78]:
from scipy.stats import t

In [79]:
t.ppf(0.05,2478)

-1.6454687772470247

In [80]:
t.ppf((1-0.025),22333)

1.9600702128845593

With the statistic method the t-statistics is bigger than the upper critical value, we accept H0

In [81]:
# Method 2: Use the p-value method.
# Your code here:
#t.cdf(-35.80631941460793,22333)

2.2589156167156547e-273

In [88]:
stat, pval = ttest_1samp(wage_trans, popmean = 35, alternative = "greater")

In [89]:
pval

5.012771694203921e-76

With the P value method,the P value is less than 0.05,so we accept the H0.

In [87]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
from scipy.stats import ttest_1samp
ttest_1samp(wage_trans,35, alternative='greater')

Ttest_1sampResult(statistic=19.064990324906386, pvalue=5.012771694203921e-76)

Accept H0

Ttest_1sampResult(statistic=19.064990324906383, pvalue=5.012771694204206e-76)

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level.

In [90]:
# Method 1: Get the critical values which correspond to a 95% confidence.
# Your code here:
wage=df2['HourlyWage']
from scipy.stats import norm
norm.interval(alpha=0.95,loc=np.mean(wage), scale=sem(wage)) 

C:\Users\mahna\AppData\Local\Temp\ipykernel_12620\3701301347.py:5: DeprecationWarning: Use of keyword argument `alpha` for method `interval` is deprecated. Use first positional argument or keyword argument `confidence` instead.
  norm.interval(alpha=0.95,loc=np.mean(wage), scale=sem(wage))


(49.79826596983879, 50.463172698540085)

In [93]:
sample_mean=np.mean(wage)
sample_mean

50.130719334189436

In [91]:
import statsmodels.stats.api as sms
sms.DescrStatsW(wage).tconfint_mean(alpha=0.05)#just another way doing it

(49.798255413084874, 50.46318325529572)

Sample mean: 50.13
Confidence interval: [49.80,50.46]


Now compute a 95% confidence interval for the hourly salary of all the Police employees.

In [94]:
# Your code here:
hour_poli= salary_poli/(50*40)
norm.interval(alpha=0.95,loc=np.mean(hour_poli), scale=sem(hour_poli)) 

C:\Users\mahna\AppData\Local\Temp\ipykernel_12620\2645842802.py:3: DeprecationWarning: Use of keyword argument `alpha` for method `interval` is deprecated. Use first positional argument or keyword argument `confidence` instead.
  norm.interval(alpha=0.95,loc=np.mean(hour_poli), scale=sem(hour_poli))


(72.9586542074945, 75.3903879177025)

In [95]:
np.mean(hour_poli)

74.1745210625985

Sample mean: 74.17
Confidence interval: [72.96,75.39]


# Chi2 test

Now we want to know if the amount of full time "FT" and part time "PT" employees is equal between Lawers, Meds, Police, Firemen and other departments. 

Considering all the options in this groups of employees will be very time consuming. To simplify this process, create first a function that returns:

* "Policemen" if "Police" is found on "JobTitle"
* "Firemen" if "Fire" is found on "JobTitle"
* "Medical" if "Med" or "Nurse" is found on "JobTitle"
* "Lawyer" if "Attorney" is found on "JobTitle"
* "Other" in any other cases

Then, create a new column named "employee_group" that determines to which group belong the employee. 

In [96]:
# Your code here: (Create the function)
def group (x):
    if ('Police' in x):
        return 'Policeman'
    elif ('Fire' in x):
        return 'Fireman'
    elif ('Med' in x):
        return 'Medical'
    elif ('Nurse' in x):
        return 'Medical'
    elif ('Attorney' in x):
        return 'Lawyer'
    else:
        return 'Other'

In [97]:
# Your code here: (Apply the function)
df2['employee_group']=df2['JobTitle'].apply(group)

In [98]:
df2.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  HourlyWage employee_group  
0  San Francisco     PT  255.366340          Other  
1  San Francisco     FT  239.826105        Medical  
2  San Francisco     PT  218.112180          Other  
3  San Francisco     FT  209.009610      Policeman  
4  San Francisco     FT  208.717550        Fireman

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage employee_group  
0  San Francisco     PT   255.366340          Other  
1  San Francisco     FT   239.826105        Medical  
2  San Francisco     PT   218.112180          Other  
3  San Francisco     FT   209.009610      Policemen  
4  San Francisco     FT   208.717550        Firemen

Determine how many "PT" and "FT" employess have all the employees groups.

In [99]:
# Your code here: (Store the output dataframe into a new variable)

#df2.groupby(['employee_group','Status']).size()
#df2.value_counts(subset=['employee_group', 'Status']) 
#df2.groupby(['employee_group'])['Status'].count()
#df2.groupby(['employee_group','Status']).agg({'Status':'count'})
#crosstab=pd.crosstab(df2['employee_group'],df2['Status'],values=df2['Status'],aggfunc='count')
crosstab = pd.crosstab(df2['employee_group'], df2['Status']) 
crosstab

Status             FT     PT
employee_group              
Fireman          1333    178
Lawyer            317    102
Medical          1028   2889
Other           18126  12245
Policeman        1530    371

Status             FT     PT
employee_group              
Firemen          1333    178
Lawyers           317    102
Medical          1028   2889
Other           18126  12245
Policemen        1530    371

Now try compute the expected frequencies doing the calculations with the individual probabilities. Remember that the Chi2 test assumes that both variables (employee_group and FT/PT) are not related (therefore they are independent). Therefore, to compute the expected frequencies you need to compute the probability of each cell and multiply it by the number of observations. ie:

$$\nu(x,y) = p(x,y) * N = p(x) * p(y) * N$$

bear in mind that in general: $p(x,y)\neq p(x)*p(y)$; the equality will only be true if x and y are independent. However, the null hypotheses says that **x and y are independent.** but that's what we're assuming with the null hypotheses.

where "x" is the "employee_group" and "y" the (FT/PT). 

In [100]:
# Create an empty dataframe named "frequencies" to store the data.
# Your code here:
from scipy.stats import chi2_contingency
chi2_contingency(crosstab)

(2676.642333711905,
 0.0,
 4,
 array([[  885.29798788,   625.70201212],
        [  245.49295627,   173.50704373],
        [ 2294.97830478,  1622.02169522],
        [17794.43096618, 12576.56903382],
        [ 1113.79978488,   787.20021512]]))

In [101]:
frequency = pd.DataFrame()

In [102]:
df2.shape

(38119, 14)

In [103]:
ft=1333+317+333+18821+1530
ft

22334

In [ ]:
# Your code here: Compute Expected frequency of being "Firemen" and "FT". Store the solution in a variable named "firemen_ft"

In [104]:
firemen_ft = ((1333+178)/38119)*(22334/38119)*38119
firemen_ft

885.2979878800596

885.2979878800596

In [ ]:
# Your code here: Compute Expected frequency of being "Firemen" and "PT". Store the solution in a variable named "firemen_pt"

In [105]:
firemen_pt = ((1333+178)/38119)*((38119-22334)/38119)*38119
firemen_pt

625.7020121199403

625.7020121199403

In [ ]:
# Your code here: Compute Expected frequency of being "Lawyers" and "FT". Store the solution in a variable named "lawyers_ft"

In [106]:
lawyers_ft = ((317+102)/38119)*(22334/38119)*38119
lawyers_ft

245.49295626852748

245.49295626852748

In [ ]:
# Your code here: Compute Expected frequency of being "Lawyers" and "PT". Store the solution in a variable named "lawyers_pt"

In [107]:
lawyers_pt = ((317+102)/38119)*((38119-22334)/38119)*38119
lawyers_pt

173.5070437314725

173.5070437314725

In [ ]:
# Your code here: Compute Expected frequency of being "Medical" and "FT". Store the solution in a variable named "medical_ft"

In [108]:
medical_ft = ((1028+2889)/38119)*(22334/38119)*38119
medical_ft

2294.9783047823917

2294.9783047823917

In [ ]:
# Your code here: Compute Expected frequency of being "Medical" and "PT". Store the solution in a variable named "medical_pt"

In [109]:
medical_pt = ((1028+2889)/38119)*((38119-22334)/38119)*38119
medical_pt

1622.021695217608

1622.021695217608

In [ ]:
# Your code here: Compute Expected frequency of being "Other" and "FT". Store the solution in a variable named "other_ft"

In [110]:
other_ft = ((18126+12245)/38119)*(22334/38119)*38119
other_ft

17794.430966184842

17794.430966184842

In [ ]:
# Your code here: Compute Expected frequency of being "Other" and "PT". Store the solution in a variable named "other_pt"

In [111]:
other_pt = ((18126+12245)/38119)*((38119-22334)/38119)*38119
other_pt

12576.569033815158

12576.569033815158

In [ ]:
# Your code here: Compute Expected frequency of being "Policement" and "FT". Store the solution in a variable named "policemen_ft"

In [113]:
policemen_ft = ((1530+371)/38119)*(22334/38119)*38119
policemen_ft

1113.7997848841785

1113.7997848841785

In [ ]:
# Your code here: Compute Expected frequency of being "Policement" and "PT". Store the solution in a variable named "policemen_pt"

In [114]:
police_pt = ((1530+371)/38119)*((38119-22334)/38119)*38119
police_pt

787.2002151158215

787.2002151158215

* Store all the expected frequencies of "FT" employees in a list 
* Store all the "PT" employees into another list
* Create a dictionary with "FT" and "PT" as keys and as the values use the previous lists
* Create a dataframe with this dictionary using pd.DataFrame()

In [115]:
# Your code here:
ft=[885.2979878800596,245.49295626852748,2294.9783047823917,17794.430966184842,1113.7997848841785]
pt=[625.7020121199403,173.5070437314725,1622.021695217608,12576.569033815158,787.2002151158215]
dic={'FT':ft,'PT':pt}
dic=pd.DataFrame(dic)
dic

FT            PT
0    885.297988    625.702012
1    245.492956    173.507044
2   2294.978305   1622.021695
3  17794.430966  12576.569034
4   1113.799785    787.200215

FT            PT
0    885.297988    625.702012
1    245.492956    173.507044
2   2294.978305   1622.021695
3  17794.430966  12576.569034
4   1113.799785    787.200215

Now use the "st.chi2_contingency()" from scipy.stats [documentation here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) to conduct a Chi2 test to determine if the diferences between employee groups are statistically significant using a 95% confidence level. Hint: fill the function with a dataframe of actual frequencies.

In [116]:
# Your code here: (use the st.chi2_contingency() function from scipy.stats to compute:
# The Chi2 value
# The p-valueYea we 
# The expected frequencies.
# Ho: there is no relationship
# Ha: there is relationship differences
# p_value = P(table | Ho) = P(table | no relationship) = 1.51e-6 < 0.05
from scipy.stats import chi2_contingency
chi2_contingency(crosstab)

(2676.642333711905,
 0.0,
 4,
 array([[  885.29798788,   625.70201212],
        [  245.49295627,   173.50704373],
        [ 2294.97830478,  1622.02169522],
        [17794.43096618, 12576.56903382],
        [ 1113.79978488,   787.20021512]]))

(2676.642333711905, 0.0, 4, array([[  885.29798788,   625.70201212],
        [  245.49295627,   173.50704373],
        [ 2294.97830478,  1622.02169522],
        [17794.43096618, 12576.56903382],
        [ 1113.79978488,   787.20021512]]))

Check if your expected frequencies aggree with the ones obtained with the st.chi2_contingency() function

Yes, they are exactly the same.